photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/NameAndTermFeatureSetContainer.scala

```diff
@@ -62,24 +64,44 @@ protected[ml] class NameAndTermFeatureSetContainer(nameAndTermFeatureSets: Map[S
   def saveAsTextFiles(nameAndTermFeatureSetContainerOutputDir: String, sc: SparkContext): Unit = {
     nameAndTermFeatureSets.foreach { case (featureSectionKey, featureSet) =>
       val featureSetPath = new Path(nameAndTermFeatureSetContainerOutputDir, featureSectionKey)
-      saveNameAndTermSetAsTextFiles(featureSet, sc, featureSetPath)
+      saveNameAndTermSetAsTextFiles(featureSet, featureSetPath, sc)
     }
   }
 
   /**
-   * Write the [[Set]] of [[NameAndTerm]]s to HDFS as text files.
+   * Write the [[RDD]] of [[NameAndTerm]]s to HDFS as a single text file.
    *
    * @param nameAndTermSet The map to be written
-   * @param sc The Spark context
    * @param outputPath The HDFS path to which write the map
+   * @param sc The Spark context
    */
   private def saveNameAndTermSetAsTextFiles(
-      nameAndTermSet: Set[NameAndTerm],
-      sc: SparkContext,
-      outputPath: Path): Unit = {
+      nameAndTermSet: RDD[NameAndTerm],
+      outputPath: Path,
+      sc: SparkContext): Unit = {
+    val hadoopFS = FileSystem.get(sc.hadoopConfiguration)
+    val tmpPath = new Path(outputPath.getParent, s"${outputPath.getName}-tmp")
+
+    // saveAsTextFile in a temporary folder
+    hadoopFS.delete(tmpPath, true)
+    // save result uncompressed for human-readable output
+    sc.hadoopConfiguration.set("mapred.output.compress", "false")
+    nameAndTermSet.map(_.toString).saveAsTextFile(tmpPath.toString)
 
-    val iterator = nameAndTermSet.iterator.map(_.toString)
-    IOUtils.writeStringsToHDFS(iterator, outputPath, sc.hadoopConfiguration, forceOverwrite = false)
+    // Merge files in temporary folder into final output
+    hadoopFS.delete(outputPath, true)
+    FileUtil.copyMerge(
+      hadoopFS, tmpPath,
+      hadoopFS, outputPath,
+      true, sc.hadoopConfiguration, null
+    )
+  }
+
+  /**
+   * Export the nameAndTermFeatureSets to a Map of keys to NameAndTerm feature sets for testing.
+   */
+  def exportAsKeyNameTermSetMap(): Map[String, Set[NameAndTerm]] = {
```

Comment:
I think functions (especially public functions) that are only used for testing purposes are an anti-pattern

Reply:
Done